docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+[source,bash]
+----
+test_init test_api_xccdf_foo.log
+----
+
+NOTE: The specified log file will contain output of both `stdout` and `stderr`
+of every executed test in a directory.
+
+
+==== test_run function
+NOTE: This function should only be called from the entry script which is
+responsible for running all the tests from a specific directory.
+
+The function is responsible for executing a test script file or a function and
+logging its result into the log file created by the `test_init` function.
+[[app-listing]]
+[source,bash]
+----
+test_run "DESCRIPTION" $srcdir/TEST_SCRIPT_FILE|TEST_FUNCTION ARG [ARG...]
+----
+The `$srcdir` variable contains the path to the directory with the entry script.
+The `test_run` function reports the following results into the log file:
+
+* *PASS* when script/function returns *0*,
+* *FAIL* when script/function returns *1*,
+* *SKIP* when script/function returns *255*,
+* *WARN* when script/function returns none of the above exit statuses.
+
+The result of every test executed by the `test_run` function will be reported
+in the log file in a following way:
+[[app-listing]]
+[source,bash]
+----
+TEST: DESCRIPTION
+<test stdout + stderr output>
+RESULT: PASS/FAIL/SKIP/WARN
+----
+
+
+==== test_exit function
+NOTE: This function should only be called from the entry script which is
+responsible for running all the tests from a specific directory.
+
+The function is responsible for cleaning-up the testing environment. You can
+call it without arguments or with one argument -- a script/function which will
+do additional clean-up tasks.
+[[app-listing]]
+[source,bash]
+----
+test_exit [CLEAN_SCRIPT|CLEAN_FUNCTION]
+----
+
+
+==== require function
+Checks if requirements are in the `$PATH`, use it as follows:
+[[app-listing]]
+[source,bash]
+----
+require 'program' || return 255
+----
+
+
+==== probecheck function
+Checks if probe exists, use it as follows:
+[[app-listing]]
+[source,bash]
+----
+probecheck 'probe' || return 255
+----
+
+
+==== verify_results function
+Verifies that there is the `COUNT` number of results of selected OVAL `TYPE` in
+a `RESULTS_FILE`:
+[[app-listing]]
+[source,bash]
+----
+verify_results TYPE CONTENT_FILE RESULTS_FILE COUNT
+----
+
+NOTE: This function expects that the OVAL `TYPE` is numbered from `1` to `COUNT`
+in the `RESULTS_FILE`.
+
+
+==== assert_exists function
+Does an XPath query to a file specified in `$result` variable and checks if
+number of results matches with an expected number specified as an argument:
+[[app-listing]]
+[source,bash]
+----
+result="relative_path_to_file"
+assert_exists EXPECTED_NUMBER_OF_RESULTS XPATH_QUERY_STRING
+----
+
+For example, let's say you want to check that in the `results.xml` file the
+result of the rule `xccdf_com.example.www_rule_test` is fail:
+[[app-listing]]
+[source,bash]
+----
+result="./results.xml"
+my_rule_="xccdf_com.example.www_rule_test"
+assert_exists 1 "//rule-result[@idref=\"$my_rule\"]/result[text()=\"fail\"]"
```

Comment:
Just a side question: Does it work with single quotes instead of escaping double quotes?

Reply:
Yes it does.